### ISSUE TITLE
Lazy-loaded frames are treated as loaded

### ISSUE THREAD

**Issue by danmcge** [hotwired/turbo#214]
Suppose you have a menu, loaded in with a frame (`src="/menu"`), that is initially off screen (with a `<details>`/`<summary>` tag pairing, say). By setting `loading="lazy"` on the frame, we can defer loading until someone opens the menu. Great. But what if you want to focus the first menu item when the menu loads?

We need to wait for the frame to load, because it will load in all our menu items. We have nothing to focus until the frame loads. As frame elements have a `loaded` promise, the solution seems really easy — a Stimulus controller action something like:

```js
async adjustFocus() {
  await this.frameElement.loaded
  this.focusFirstItem()
}
```

This won’t work, though. Frame elements are [instantiated with the promise resolved](https://github.com/hotwired/turbo/blob/8bce5f17cd697716600d3b34836365ebcdc04b3f/src/elements/frame_element.ts#L18). And the only time the promise gets changed to a pending state is [when we load the frame src](https://github.com/hotwired/turbo/blob/main/src/core/frames/frame_controller.ts#L66). But because we’re lazy loading, we don’t load the frame src until we open the menu, so when the `adjustFocus` method above runs, the promise is already resolved. It doesn’t need to `await` the promise, it just goes on to `focusFirstItem`. Yet the frame hasn’t loaded, we have no items, so this is going to fail.

**Comment by danmcge** [comment]
I think the solution here is something along the lines of: (1) assign a thenable object to `this.element.loaded` in FrameController#connect and (2) resolve that "thenable", as a promise, in FrameController#elementAppearedInViewport. But I’m too much of a typescript noob to make it work.

This is a first draft of the implementation I had in mind:
```js
//  src/core/frames/frame_controller.ts

export class FrameController {
  // ...

  connect() {
    if (this.loadingStyle == FrameLoadingStyle.lazy) {
      this.element.loaded = { then: (resolve) => resolve() }
      this.appearanceObserver.start()
    }
    
    // ...
  }
  
  // ...
  
  async elementAppearedInViewport(element: Element) {
    const promise = this.element.loaded
    await this.loadSourceURL()
    await Promise.resolve(promise)
  }
}
```

**Comment by sstephenson** [comment]
Agree, you should be able to wait on the loaded promise for a lazy frame.

**Comment by mattalat** [comment]
I believe this is still an issue. There's also the question of why lazy frames have fulfilled `.loaded` promises and are `.complete == true` before they're visible on the screen.

**Comment by bellef** [comment]
Upvoting as it's still an issue 👍

**Comment by tvongaza** [comment]
Also up voting, just ran into this bug.

**Comment by tpaulshippy** [comment]
Looking at putting together a fix [here](https://github.com/hotwired/turbo/compare/main...tpaulshippy:turbo:issue-214). Hit an issue with an existing test, but I can take another look as soon as I can.
